Add a client for sending stats to prometheus

tiyash-basu-sociomantic opened the pull request:
Prometheus sends HTTP GET requests to an application at `/metrics` endpoint to pull stats from it. This PR adds a listener for these requests, and a handler to response with stats to prometheus.

Also accompanying here are classes that interface between the handler and a client application, such that the handler is able to invoke callbacks that collect stats from the client application upon receiving requests from prometheus.

tiyash-basu-sociomantic committed:
Add modules to collect and format stats for Prometheus

While pulling stats from clients, Prometheus expects the stats to be
formatted in a specific way. The modules added in this commit are able to
read stats (metrics and labels) from structs or classes via a callback
mechanism, and format them in way that can be accepted by Prometheus.

To know more about Prometheus' stats format, please refer to
`https://prometheus.io/docs/instrumenting/exposition_formats/`.

To know more about Prometheus' data model, and specific terminologies like
Metrics and Labels, please refer to
`https://prometheus.io/docs/concepts/data_model/`

tiyash-basu-sociomantic committed:
Add missing libraries to test targets

tiyash-basu-sociomantic committed:
Add an integration test for stat collection with prometheus

tiyash-basu-sociomantic committed:
Add prometheus stat collection delegates to DaemonApp

These delegates can be sent to the the CollectorRegistry instance used to
instantiate the prometheus request listener in the client application.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/server/PrometheusListener.d:
```diff
@@ -0,0 +1,161 @@
+
+    Derives from `SelectListener` with it's handler specified as
+    `PrometheusHandler`.
+
```
Could you document the template params?

gavin-norman-sociomantic replied:
Oh wait, sorry, I misread this. I thought that these were params of `PrometheusListener`, whereas they're actually params used to instantiate `SelectListener`. Probably better to remove the comments then. Sorry!

gavin-norman-sociomantic commented on src/ocean/util/prometheus/server/PrometheusListener.d:
```diff
@@ -0,0 +1,161 @@
+    stats and labels, the callbacks will be called for every incoming prometheus
+    request, and the accumulated stats will be appended to the response message.
+
+    Example:
```
Would this work as a documented unittest?

gavin-norman-sociomantic commented on src/ocean/util/prometheus/server/PrometheusHandler.d:
```diff
@@ -0,0 +1,155 @@
+
+*******************************************************************************/
+
+public class PrometheusHandler : HttpConnectionHandler
```
Ah, shame we don't have an `HttpTaskConnectionHandler` yet, but never mind.

tiyash-basu-sociomantic replied:
We do have a `ocean.net.http.TaskHttpConnectionHandler`, which makes using the task system (or at least the global `theScheduler` instance) mandatory. I thought using `HttpConnectionHandler` would be slightly less restrictive, while offering the same set of functionalities.

Do you suggest that we use `TaskHttpConnectionHandler` here?

gavin-norman-sociomantic replied:
Ha! I was looking in the wrong place for it ;)

Yeah, on the whole I agree that compatibility is probably more important here.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/server/PrometheusHandler.d:
```diff
@@ -0,0 +1,155 @@
+    An HTTP response handler for prometheus' requests, used in
+    `PrometheusListener`
+
+    Responds to GET requests on;y..
```
Typo.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/server/PrometheusHandler.d:
```diff
@@ -0,0 +1,155 @@
+    {
+        if (this.request.uri_string() != "/metrics")
+        {
+            PrometheusHandler.log.info("Received unexpected request uri: {}",
```
Warning?

tiyash-basu-sociomantic replied:
I will update the message being logged here to be more specific.

Prometheus only needs the `/metrics` endpoint for pulling stats, and hence every other endpoint results in a `NotImplemented` response here. However, this is a general HTTP handler listening at a port, so anyone can send it a GET request at any endpoint. So, in that sense, perhaps it is not necessarily a warning?

gavin-norman-sociomantic commented on src/ocean/util/prometheus/server/PrometheusHandler.d:
```diff
@@ -0,0 +1,155 @@
+        }
+        catch (Exception ex)
+        {
+            err_buf.length = 0;
```
Also log the exception?

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/CollectorRegistry.d:
```diff
@@ -0,0 +1,209 @@
+
+module ocean.util.prometheus.collector.CollectorRegistry;
+
+/*******************************************************************************
```
You can use `/// ditto` here.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/CollectorRegistry.d:
```diff
@@ -0,0 +1,209 @@
+        and returns the accumulated result.
+
+        Returns:
+            The stats accumulated since last time this method was called.
```
Worth noting what format these are in? (e.g. mention the module that does the rendering / link to online doc about the format?)

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/CollectorRegistry.d:
```diff
@@ -0,0 +1,209 @@
+    }
+}
+
+/// Test collection from a single delegate
```
Are these intended as tests or usage examples? (They look like tests to me, as I don't suppose using `toDg` is recommended usage.) Any unittest commented with `///` will be treated as a usage example, so you might want to just use `//` for these.

tiyash-basu-sociomantic replied:
I removed usages of `toDg` from these tests (it looks like it was needed before because the calbacks were module-level functions). If it's alright, I have kept the tests commented with `///` to make them serve as examples as well.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
```diff
@@ -0,0 +1,332 @@
+    (https://prometheus.io/docs/concepts/metric_types/) have not been
+    implemented yet. So, as of now, this collector can can process only
+    primitive data-type members from a struct or a class. However, the current
+    prometheus stat collection framework can support metrics with very little
```
"can support" -> "could be enhanced to support" ?

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
```diff
@@ -0,0 +1,332 @@
+
+        void collect ( Collector collector )
+        {
+            collector.collect(this.system_stats.collect());
```
Maybe add a comment: `CpuMemoryStats.collect() returns a struct whose fields we want to collect via Prometheus.`

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
````diff
@@ -0,0 +1,332 @@
+        ```
+
+    2.  For collecting stats from a struct or class, and with one label:
+        `void collect ( istring LabelName, ValuesT, LabelT ) (
````
You need to explain how the label works. It's not clear to me, as someone who's not read anything about Prometheus before.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
````diff
@@ -0,0 +1,332 @@
+        }
+        ```
+
+    3.  For collecting stats from a struct or class, along with a set of labels
````
Ditto about the labels.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
```diff
@@ -0,0 +1,332 @@
+
+    This class provides three methods for stat collection:
+
+    1.  For collecting stats from a struct or class:
```
Could these examples be documented unittests? (Verbal explanations here, code examples as documented unittests?)

tiyash-basu-sociomantic replied:
The examples there were using `CpuMemoryStats.collect()` method, which would return dynamic values, which may make it a bit dodgy for unit tests. A similar example now exists as a documented unit test in `PrometheusListener`.

This module already had some unit tests (the ones you mentioned about [here](https://github.com/sociomantic-tsunami/ocean/pull/699#discussion_r278489075)). Now only some act as documentation/examples (the rest are now only test-cases) which roughly cover the use-cases of in this class.

gavin-norman-sociomantic replied:
The reason for avoiding example code in comments is that it's not being compiled, so is very vulnerable to getting out-dated.

tiyash-basu-sociomantic replied:
Right, I agree with you. I also noticed, that although I removed the examples, I forgot to remove the method signatures. :)

Edit: Removed the method signatures as well.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
```diff
@@ -0,0 +1,332 @@
+    /// method is called.
+    private mstring collect_buf;
+
+    /// Default constructor.
```
I don't think there's any need to include this.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
```diff
@@ -0,0 +1,332 @@
+    /***************************************************************************
+
+        Returns:
+            The stats collected until the last call to the `reset` method, in a
```
"until" -> "since".

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
```diff
@@ -0,0 +1,332 @@
+        Returns:
+            The stats collected until the last call to the `reset` method, in a
+            textual respresentation that can be readily added to a response
+            message body.
```
Would also be nice to link something that describes the format here.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/Collector.d:
```diff
@@ -0,0 +1,332 @@
+    }
+}
+
+/// Test collecting stats having initial values, but without any label.
```
Looks like these are tests, not usage examples, so should have `//` instead of `///`.

gavin-norman-sociomantic commented on src/ocean/util/prometheus/collector/StatFormatter.d:
```diff
@@ -0,0 +1,384 @@
+
+*******************************************************************************/
+
+public final abstract class StatFormatter
```
More idiomatic would be to not use a class and instead declare your functions at module scope. If anyone wants to wrap them in a "namespace", they can do so with a named import.

gavin-norman-sociomantic reviewed (commented):
Some minor comments, but generally this is looking really good. Nicely structured. Good work!

The biggest thing, for me, is the lack of description about some of the Prometheus terminology, e.g. the labels in the `format` functions.

tiyash-basu-sociomantic committed:
Add an HTTP server to handle Prometheus scrapes

Prometheus sends HTTP GET requests to an application at `/metrics` endpoint
to scrape stats from it. This commits adds a listener for these requests,
and a handler to response with stats to Prometheus.

The listener needs to be instantiated by the application, with a set of
callbacks, via the `CollectorRegistry` module, that can be called by the
handler to prepare the response messages.

tiyash-basu-sociomantic commented:
Thanks for the review!

I have now added some additional documentation about the Prometheus-specific terminology used here. I have also included links to relevant pages on Prometheus' website as references.

Ready for another look.

tiyash-basu-sociomantic committed:
Create `.keep` in relnotes directory

This helps the directory to appear after pulling, when there are no release
notes there.

tiyash-basu-sociomantic committed:
Add a release note for prometheus integration

tiyash-basu-sociomantic commented:
Created a `.keep` file in the `relnotes` folder, and added a release note for the prometheus integration feature.

gavin-norman-sociomantic reviewed (commented):
New doc greatly improves comprehensibility :+1:

tiyash-basu-sociomantic commented:
I have removed the documentation as you pointed out.

Thanks for the review!
